packages/common-all/src/notePropDict.ts

```diff
@@ -0,0 +1,134 @@
+import _ from "lodash";
+import {
+  NotePropsDict,
+  NotePropsByFnameDict,
+  DVault,
+  NoteProps,
+  NotePropsFullDict,
+} from "./types";
+import { cleanName, isNotUndefined } from "./utils";
+import { VaultUtils } from "./vault";
+
+/**
+ * Utilities for working with NotePropsFullDict. The reason NotePropsFullDict is not a class is due to needing
+ * to work with primitive objects with redux
+ */
+export class NoteFullDictUtils {
+  /**
+   * Find notes by fname. If vault is provided, filter results so that only notes with matching vault is returned
+   * Return empty array if no match is found
+   *
+   * @param fname
+   * @param notesDict
+   * @param vault If provided, use to filter results
+   * @returns Copy of NoteProps array
+   */
+  static findNotesByFname(
+    fname: string,
+    notesDict: NotePropsFullDict,
+    vault?: DVault
+  ): NoteProps[] {
+    const { notesById, notesByFname } = notesDict;
+    const cleanedFname = cleanName(fname);
+    const ids: string[] | undefined = notesByFname[cleanedFname];
+    if (!ids) {
+      return [];
+    }
+    let notes = ids.map((id) => notesById[id]).filter(isNotUndefined);
+    if (vault) {
+      notes = notes.filter((note) => VaultUtils.isEqualV2(note.vault, vault));
+    }
+    return _.cloneDeep(notes);
+  }
+
+  /**
+   * Remove note from both notesById and notesByFname
+   *
+   * @param note note to delete
+   * @param notesDict
+   */
+  static deleteNote(note: NoteProps, notesDict: NotePropsFullDict) {
```

Comment:
should we return an error if the key isn't present?

Reply:
Looks like its fairly standard to return a boolean instead to denote whether it was deleted or it. I can return that instead. We can let the clients construct the error (eg 404) if needed